system/security/plugins/testauthSecurity/testauthSecurity.cpp

```diff
@@ -0,0 +1,591 @@
+        Owned<CAuthMap> authmap = new CAuthMap();
+
+        Owned<IPropertyTreeIterator> iter = authconfig->getElements(".//Location");
+        ForEach(*iter)
+        {
+            IPropertyTree& location = iter->query();
+            StringBuffer pathStr, rstr, required, description;
+            location.getProp("@path", pathStr);
+            location.getProp("@resource", rstr);
+            location.getProp("@required", required);
+            location.getProp("@description", description);
+
+            if (pathStr.length() == 0)
+                throw makeStringException(-1, "path empty in Authenticate/Location");
+            if (rstr.length() == 0)
+                throw makeStringException(-1, "resource empty in Authenticate/Location");
+
+            ISecResourceList* rlist = authmap->queryResourceList(pathStr);
+            if (rlist == nullptr)
+            {
+                rlist = createResourceList("htpasswdsecurity", secureContext);
+                authmap->add(pathStr, rlist);
+            }
+            ISecResource* rs = rlist->addResource(rstr);
+            SecAccessFlags requiredaccess = str2perm(required);
+            rs->setRequiredAccessFlags(requiredaccess);
+            rs->setDescription(description);
+        }
+        return authmap.getClear();
+    }
+
+    IAuthMap* createFeatureMap(IPropertyTree* authconfig, IEspSecureContext* secureContext = nullptr) override
+    {
+        Owned<CAuthMap> authmap = new CAuthMap();
+
+        Owned<IPropertyTreeIterator> iter = authconfig->getElements(".//Feature");
+        ForEach(*iter)
+        {
+            IPropertyTree& feature = iter->query();
+            StringBuffer pathStr, rstr, required, description;
+            feature.getProp("@path", pathStr);
+            feature.getProp("@resource", rstr);
+            feature.getProp("@required", required);
+            feature.getProp("@description", description);
+            ISecResourceList* rlist = authmap->queryResourceList(pathStr);
+            if(rlist == nullptr)
+            {
+                rlist = createResourceList(pathStr, secureContext);
+                authmap->add(pathStr, rlist);
+            }
+            if (!rstr.isEmpty())
+            {
+                ISecResource* rs = rlist->addResource(rstr);
+                SecAccessFlags requiredaccess = str2perm(required);
+                rs->setRequiredAccessFlags(requiredaccess);
+                rs->setDescription(description);
+            }
+        }
+        return authmap.getClear();
+    }
+
+    IAuthMap* createSettingMap(IPropertyTree* authConfig, IEspSecureContext* secureContext = nullptr) override
+    {
+        return nullptr;
+    }
+
+    SecAccessFlags getAccessFlag(SecResourceType rtype, ISecUser& user, const char* resourceName, IEspSecureContext* secureContext)
+    {
+        SecAccessFlags  resultFlag = SecAccess_Unavailable;
+        try
+        {
+            if (!authenticate(user))
+                return resultFlag;
```

Comment:
minor: Clearer when reading/reviewing the code if this (and line 423) return SecAccess_Unavailable.

Reply:
Will change